### PR TITLE
ENH: enable to override all default_units with units_override (Pluto)

### PR DESCRIPTION
**Pull request by xshaokun**
### PR Summary

- enable to derive all default_units with the units passed to `units_override`
- adds test for `units_override`

### CODE REVIEW

File: yt_idefix/data_structures.py
Comment (neutrinoceros) [inline yt_idefix/data_structures.py:449]: please add a comment here to note that `"density_unit"` is the one key that's actually not in the base `Dataset` class

File: yt_idefix/data_structures.py
Comment (neutrinoceros) [inline yt_idefix/data_structures.py:471]: Ah so if I'm not mistaken, you don't actually need to override this method, as you would get the same result by just changing `IdefixVtkDataset.default_units` to only have the three keys you consider valid here `("length_unit", "velocity_unit", "density_unit")`. However, this is probably not a valid approach because then we wouldn't have a default for magnetic and temperature units.

Most importantly, your current implementation is problematic because it prevents using otherwise valid unit normalisation (in `mass_unit` should still be usable).

I believe I designed (or refactored, rather) this class method in yt a while ago. I believe my intention was to allow _extending_ the list of valid units overrides, but it's probably not a good idea to restrict it, compared to the base class.

**Review by neutrinoceros — CHANGES_REQUESTED** [review]
While I can certainly get behind what you want to enable here, I think the current implementation isn't backward compatible, but we can certainly evolve it to be !

Also I think this count as a feature rather than a bugfix, so you could just bump the version number to 0.10.0

File: yt_idefix/data_structures.py
Comment (xshaokun) [reply yt_idefix/data_structures.py:471]: Your point is right. I found the method you designed in the AMRVAC frontend, that's why I commented here that other combinations would be supported in the future. 

I planed to make units_override usable for now. Maybe I don't have to rush to create this PR, so I'm considerring putting this PR on hold until I make this to be a mature feature.

Comment (xshaokun) [reply yt_idefix/data_structures.py:471]: My intention is to reduce the dependece on header file `definitions.h` during analysis, and be able to scale the results to other physical system by adapting normalization factor in post-analysis.

Comment (neutrinoceros) [reply yt_idefix/data_structures.py:471]: >  I'm considerring putting this PR on hold until I make this to be a mature feature.

That is fine by me.

> My intention is to reduce the dependece on header file definitions.h during analysis, and be able to scale the results to other physical system by adapting normalization factor in post-analysis.

Sounds like a reasonable approach.

**Comment by xshaokun** [comment]
Oops I have a difficulty now and would like to share some thoughts with you. If you have any advice, that would be great.

the `default_unit` now consists seven units: length, time, velocity, density, magnetic and temperatuer. In Pluto, [length, density, velocity] are used as base units and are used to derived other units. But in post-analysis, beside temperature should not be override since it's always in Kelvin in Pluto, there's no such limitation and users are able to derive all units based on any combinations of three units other than [length, velocity, time]. 

The difficulty is there are too much combinations! The worst method is to use many if/else statements to derive other units based on different combinations, but that must be terrible. So I'm trying to find some constraints to reduce valid combinations.

**Comment by neutrinoceros** [comment]
> The difficulty is there are too much combinations! The worst method is to use many if/else statements to derive other units based on different combinations, but that must be terrible. So I'm trying to find some constraints to reduce valid combinations.

yeah that sounds difficult to get right and maintain. I am not sure how I would approach this but I would advise you consider generating the list of valid combinations instead of hard-coding it. I'm thinking it could be relatively easy using `itertools` functions or possibly `more_itertools` (third party, but already a hard dependency to yt it self, so a transitive dependency to yt_idefix).

**Comment by xshaokun** [comment]
> I'm thinking it could be relatively easy using itertools functions or possibly more_itertools

Thanks! I”ll take a look tomorrow.

**Comment by xshaokun** [comment]
Many condition statements cannot be avoided for now, but it's much better than I expected. I'll come back tomorrow if you have any comments :)

**Comment by neutrinoceros** [comment]
I cannot promise to review this before next Wednesday, but I'll try.

**Comment by xshaokun** [comment]
It's okay，take your time

File: tests/test_vtk.py
Comment (neutrinoceros) [inline tests/test_vtk.py:60]: please import the  PlutoVtkDataset class at the top level. It'll be easier to migrate this frontend to yt's main repo if every name changes are gathered at the top level of each files (or as much as possible)
```suggestion
    invalid = PlutoVtkDataset.invalid_unit_combinations
```

File: tests/test_vtk.py
Comment (neutrinoceros) [inline tests/test_vtk.py:58]: Can you add a comment to explain how you chose these values (randomly ? or is this a specific set of consistent values ?)

File: tests/test_vtk.py
Comment (neutrinoceros) [inline tests/test_vtk.py:66]: I'd like the error message to be even clearer. Please explicitly say that the input units_override isn't valid

File: yt_idefix/data_structures.py
Comment (neutrinoceros) [inline yt_idefix/data_structures.py:431]:
```suggestion
        # User-passed units are preferred to PLUTO's base units
```

**Review by neutrinoceros — COMMENTED** [review]
I didn't have time to look at everything but here are a couple comments already. I'll finish reading this soon

File: tests/test_vtk.py
Comment (xshaokun) [reply tests/test_vtk.py:60]: It makes sense!

File: tests/test_vtk.py
Comment (xshaokun) [reply tests/test_vtk.py:58]: Sure. I just chose the first three values arbitraily and derived others to make them consistent.

File: tests/test_vtk.py
Comment (xshaokun) [reply tests/test_vtk.py:66]: Okay, I'll improve it

File: yt_idefix/data_structures.py
Comment (xshaokun) [inline yt_idefix/data_structures.py:599]: @neutrinoceros Since there are too many valid combinations, I only listed invalid combinations here to let users avoid. Should I explicitly show this verbose message in the test?

Comment (neutrinoceros) [reply yt_idefix/data_structures.py:599]: You approach seems reasonable, but yes, please include a test to coverage this path too.

File: yt_idefix/data_structures.py
Comment (neutrinoceros) [inline yt_idefix/data_structures.py:586]: Are you sure that's correct ? I would assume that overriding magnetic as well as length, mass and time is fine ?

File: yt_idefix/data_structures.py
Comment (neutrinoceros) [inline yt_idefix/data_structures.py:581]: I'm not 100% sure this is fully consistent with your inline comment above that states that user-passed overrides are preferred over Pluto's internal units. I could easily be convinced that it's fine, however.

File: yt_idefix/data_structures.py
Comment (neutrinoceros) [inline yt_idefix/data_structures.py:439]: equivalent but clearer (in my opinion)
```suggestion
            if len(self.units_override) > 0:
```

File: yt_idefix/data_structures.py
Comment (neutrinoceros) [inline yt_idefix/data_structures.py:441]:
```suggestion
                    "Less than 3 units were specified in units_override (got %s). "
```

File: yt_idefix/data_structures.py
Comment (xshaokun) [reply yt_idefix/data_structures.py:586]: Since we have only three base units here: length, mass and time, just like solving a system of equations containing three unknowns, three equations are enough, or three base/compound units here.

So an additional magnetic might conflict with that derived with length, mass and time. "more will _likely_ cause conflict" is more precise I think.

File: yt_idefix/data_structures.py
Comment (xshaokun) [reply yt_idefix/data_structures.py:446]: > note however that I don't think `self.units_override` should be updated internally. Because this attribute is user-facing, it should only reflect exactly what the user passed at load time, so I'd advise using a different intermediate variable here

I did considered that! But I also want to let users know from which three units the others are derived when they only pass 1 or 2 units, even though the additional units picked from Pluto definitions don't be overrided really. Maybe I should create another attribute?

**Comment by xshaokun** [comment]
I think I should explain more about my strategy.

The default case is no value paased to `units_override`, then the three units (density, velocity, length) from Pluto's definition are used to derived others.

In Pluto, units of density, velocity and length are unity if user doesb't specify. The `units_override` should be used when users wanted to override the (Pluto's) default units.

If the user only passes 1 unit (or two) which is not able to derive othes, so it requires two additional units. Obviously We should choose default units from PLUTO's definiton which they are consistent with valid combination.

**Comment by neutrinoceros** [comment]
Okay that makes sense. Can you also explain how magnetic units are derived here ? I don't understand how they are computed from length time and mass but without any electric charge unit.

File: yt_idefix/data_structures.py
Comment (xshaokun) [inline yt_idefix/data_structures.py:546]: @neutrinoceros Here's formula to derive magnetic, where density and velocity can be replaced with mass, time and length. There's no need of charge here because in cgs it's defined by mass, time and length

File: yt_idefix/data_structures.py
Comment (xshaokun) [reply yt_idefix/data_structures.py:599]: It's already covered in the test.

File: tests/test_vtk.py
Comment (xshaokun) [reply tests/test_vtk.py:66]: Here I just wanted to match the error messeage to make sure it's the `ValueError` I expected, not caused by something else. I think the regular expression is enough, right?

File: yt_idefix/data_structures.py
Comment (xshaokun) [reply yt_idefix/data_structures.py:546]: Honestly speaking, I didn't derive the formula by myself but it's used in Pluto and also in other frontends such as athena and amrvac.

Comment (xshaokun) [reply yt_idefix/data_structures.py:546]: In cgs, permittivity = permeability = 1 (dimensionless)

Comment (neutrinoceros) [reply yt_idefix/data_structures.py:546]: Fair enough. I'll admit I've never sat through this derivation so I don't have a grasp of what "cgs" really means in the MHD simulation realm. I'll just trust you here as I did with other collaborators and frontends :)

**Comment by neutrinoceros** [comment]
I'm planning to review this fully tomorrow unless you still want to pour more changes before I do.

File: tests/test_vtk.py
Comment (neutrinoceros) [inline tests/test_vtk.py:78]:
```suggestion
    for comb in unit_combs:
        uo = {}
        for unit in comb:
            uo[unit] = su[unit]
        if set(uo) in PlutoVtkDataset.invalid_unit_combinations:
            return
        ds = yt_idefix.load(
            pluto_vtk_file["path"], geometry=pluto_vtk_file["geometry"], units_override=uo
        )
        assert_allclose_units(ds.length_unit, ds.quan(*su["length_unit"]))
        assert_allclose_units(ds.mass_unit, ds.quan(*su["mass_unit"]))
        assert_allclose_units(ds.time_unit, ds.quan(*su["time_unit"]))
        assert_allclose_units(ds.density_unit, ds.quan(*su["density_unit"]))
        assert_allclose_units(ds.velocity_unit, ds.quan(*su["velocity_unit"]))
        assert_allclose_units(ds.magnetic_unit, ds.quan(*su["magnetic_unit"]))
```

File: tests/test_vtk.py
Comment (neutrinoceros) [inline tests/test_vtk.py:84]:
```suggestion
    ds = yt_idefix.load(pluto_vtk_file["path"], geometry=pluto_vtk_file["geometry"], units_override=uo)
```

File: tests/test_vtk.py
Comment (neutrinoceros) [inline tests/test_vtk.py:104]:
```suggestion
                pluto_vtk_file["path"], geometry=pluto_vtk_file["geometry"], units_override=uo
```

File: tests/test_vtk.py
Comment (neutrinoceros) [inline tests/test_vtk.py:92]:
```suggestion
    ds = yt_idefix.load(pluto_vtk_file["path"], geometry=pluto_vtk_file["geometry"], units_override=uo)
```

File: tests/test_vtk.py
Comment (neutrinoceros) [inline tests/test_vtk.py:131]: please avoid complicated escape sequences (`\\`) and use `re.escape` with raw strings instead.

File: tests/test_vtk.py
Comment (neutrinoceros) [inline tests/test_vtk.py:135]:
```suggestion
            pluto_vtk_file["path"], geometry=pluto_vtk_file["geometry"], units_override=SAMPLE_UNITS
```

File: yt_idefix/data_structures.py
Comment (neutrinoceros) [inline yt_idefix/data_structures.py:431]:
```suggestion
        # User-passed units are preferred over PLUTO's base units for deriving all units
```

File: yt_idefix/data_structures.py
Comment (neutrinoceros) [inline yt_idefix/data_structures.py:446]: for extra clarity
```suggestion
            # Remove those already in self.units_override
```

File: yt_idefix/data_structures.py
Comment (neutrinoceros) [inline yt_idefix/data_structures.py:448]: this loop looks a lot like a `while len(uo) < 3`. In any case I'm pretty sure it's more complex than it needs to be, so please try to simplify this a bit.

File: yt_idefix/data_structures.py
Comment (neutrinoceros) [inline yt_idefix/data_structures.py:461]:
```suggestion
        # However they are fragile, one must be cautious to change them
```

File: yt_idefix/data_structures.py
Comment (neutrinoceros) [inline yt_idefix/data_structures.py:462]: The ways these functions are designed is not immediately obvious to me. I guess I'd prefer having them named like "_setup_time_unit" ? Also the fact that they have only side-effects (they modify a variable from the outer scope rather than returning values) is a little unsettling. Can you please add some more comments to explain your strategy ?

File: yt_idefix/data_structures.py
Comment (neutrinoceros) [inline yt_idefix/data_structures.py:464]: I guess we could lower the indentation level by using guard conditions instead, such as
```suggestion
            if "time_unit" in uo:
                return
            uo["time_unit"] = uo["length_unit"] / uo["velocity_unit"]
```

**Review by neutrinoceros — COMMENTED** [review]
Alright I've read everything and I have two overarching remarks:
- I'd like to see some simplifications done before I'm confortable merging this (remember that one must be twice as smart to debug code than to write it, and this is pretty clever code already ;) )
- I'm really hoping I can count on your help and guidance in the future if we find new bugs here.

In short, I'm okay with everything you're adding here, but I'm not yet confortable maintaining it on my own. I'm sure we can make a couple things a little simpler still :)

File: yt_idefix/data_structures.py
Comment (xshaokun) [reply yt_idefix/data_structures.py:448]: Thanks! Just forgot `while`...

File: yt_idefix/data_structures.py
Comment (xshaokun) [reply yt_idefix/data_structures.py:462]: Sure.

Comment (xshaokun) [reply yt_idefix/data_structures.py:462]: > Can you please add some more comments to explain your strategy ?

Do you mean the strategy why did I set such condition statements?

I just tried different conditions during the debugging. So I'm sorry to tell you that I don't have a straightforward strategy to constrcut this part. But each one of them is useful if you test it with all the valid combinations.

**Comment by xshaokun** [comment]
Thanks for your reviewing! I did some improvements as your suggestions.

Sometimes I just don't know which kind of code is better, but I'm still learning:)

> I'm really hoping I can count on your help and guidance in the future if we find new bugs here.

Sure, I'm glad to!

File: yt_idefix/data_structures.py
Comment (neutrinoceros) [inline yt_idefix/data_structures.py:476]: so it looks like we slightly misunderstood each other (no big deal !). I suggested to rename these functions as `_setup_<x>_unit` when they were designed around side effects on an outer-scope dictionary. Now that you evolved the design to return actual values instead, a more appropriate naming convention should be `_get_<x>_unit`. I'm sorry if my original comment caused confusion.

I am not opposing your original design btw. It's fine to use side effects if you want, especially in nested functions like these.

I'm thinking however this setup function, and the myriad of nested functions you defined, look more and more like a separate, tiny class with a very specific responsibility. What do you think ?

File: yt_idefix/data_structures.py
Comment (neutrinoceros) [reply yt_idefix/data_structures.py:623]: Oh, my bad, thanks !

**Review by neutrinoceros — COMMENTED** [review]
Another round of comments. I'll aim to do a final review by Friday, and most likely merge + release then. Thanks again for your continued effort !

File: tests/test_vtk.py
Comment (xshaokun) [reply tests/test_vtk.py:81]: Of course! How could I forget it!

File: yt_idefix/data_structures.py
Comment (xshaokun) [reply yt_idefix/data_structures.py:476]: Oh it’s my first time seeing the term “side effects” in programming, I thought you mean it's bad. Now I know its real meaning.

Due to their similar function names, I wanted to use `eval` instead of a dictionary here. But I remembered you said that we should avoid using `eval`.

> 'm thinking however this setup function, and the myriad of nested functions you defined, look more and more like a separate, tiny class with a very specific responsibility. What do you think ?

For me, it's like a `unyt.UnitSystem` which however only accepts mass, length, and time. So if you recommend, I'm willing to separate this part and wrap it into a class tomorrow. I’ve been thinking it's a little bit too much to define it in this method.

Comment (neutrinoceros) [reply yt_idefix/data_structures.py:476]: Sure. I'm not familiar with how extensible unyt.UnitSystem is but if you think it's appropriate you could maybe subclass it. Otherwise just move most of this function into a (private) class and call it a day :)

File: yt_idefix/data_structures.py
Comment (neutrinoceros) [reply yt_idefix/data_structures.py:471]: We'll ultimately you're the only user I know of for now so if you don't mind having them as info it's really up to you. I'm also fine with changing the level in a future PR too if you change your mind :)

File: tests/test_vtk.py
Comment (xshaokun) [reply tests/test_vtk.py:90]:
```
expect_velocity = plut_vtk_file["units"]["length"] / ds.quan(*uo["time_unit"])
```
Maybe this is better. My intention is to check if right Pluto's unit was preserved and participated in unit derivation. The value of `expect_velocity` is different for each dataset because they have different defined units. It also depends on the unit in `uo` because if we give `length_unit` in `uo`, `density_unit` and `velocity_unit` will be preserved instead of `density_unit` and `length_unit` here. So I don't think it's a good idea to hardcode the expected value in `meta.yaml`.

Comment (neutrinoceros) [reply tests/test_vtk.py:90]: Ok that makes sense

File: yt_idefix/data_structures.py
Comment (xshaokun) [reply yt_idefix/data_structures.py:471]: I'm really hoping there'll be more Pluto's user using this frontend in the future haha.

Comment (neutrinoceros) [reply yt_idefix/data_structures.py:471]: Of course, me too !

File: yt_idefix/data_structures.py
Comment (neutrinoceros) [inline yt_idefix/data_structures.py:542]: I'm surprised you wrote this as a _nested_ class. I'd much rather have this class be seperated, otherwise I don't see the point. Also, a double trailing underscore is not a convention I'm familiar with. I think a single one is enough to indicate the class is private, isn't it ?

Comment (xshaokun) [reply yt_idefix/data_structures.py:542]: Oh when I started writing this class, I wanted to include deriving other units as well. So it shouldn't work for other code and I considered it as a nested class. Now it becomes more general but I forgot to seperate it.

Do you think I should leave this class in this file, or move it to `definitions.py` and import it here? I prefer the latter.

Comment (neutrinoceros) [reply yt_idefix/data_structures.py:542]: It doesn't need to be usable anywhere else to be separated. Otherwise you can leave it in the same file or move it to definitions if you think that's cleaner

Comment (xshaokun) [reply yt_idefix/data_structures.py:542]: Well I decide to leave this work to the future when necessary.

**Comment by xshaokun** [comment]
I'll push a new commit if you have no more suggustion :)

**Comment by neutrinoceros** [comment]
Go ahead. I'm not sure we've settled the nested-class thread though.

**Comment by xshaokun** [comment]
> I'm not sure we've settled the nested-class thread though.

I think there are no more questions (except for API for validation) on my side for now. I’ll come back tomorrow if you have any feedback.

**Review by neutrinoceros — APPROVED** [review]
Alright I see a couple minor nits I want to fix still but I'll do that myself in a follow up PR and will release in a couple hours.
Thanks again for your work, I really appreciate it !

**Comment by xshaokun** [comment]
I was going to squash my commits before merging. Anyway thanks for your patient reviews, next time I'll do it better !

**Comment by neutrinoceros** [comment]
No problem. I could have squashed on merge too but forgot to, it's no big deal :)

**Pull request merged**
